File: Volume_002/0202_At_Boss's_Expense.cpp
#include <iostream>

using namespace std;

bool p[1000001];

void primes() {
    for (int i = 0; i < 1000001; i++) {
        p[i] = true;
    }
    
    p[0] = p[1] = false;
    
    for (int i = 2; i < 1000001; i++) {
        if (p[i]) {
            for (int j = i * 2; j < 1000001; j += i) {
                p[j] = false;
            }
        }
    }
}

int main() {
    int n, x;
    primes();

    while (true) {
        cin >> n >> x;
        if (!n && !x) break;
        
        bool dp[1000001] = {false};
        int dishes[30];

        for (int i = 0; i < n; i++) {
            cin >> dishes[i];
            dp[dishes[i]] = true;
        }

        for (int i = 0; i < 1000001; i++) {
            for (int j = 0; j < n; j++) {
                if (i - dishes[j] >= 0 && dp[i - dishes[j]]) {
                    dp[i] = true;
                    break;
                }
            } 
        }

        int max = 0;
        for (int i = 0; i <= x; i++) {
            if (p[i] && dp[i]) max = i;
        }

        if (max) {
            cout << max << endl;
        } else {
            cout << "NA" << endl;
        }
        
    }

    return 0;
}

File: Volume_002/0202_At_Boss's_Expense.rb
def primes(max)
  nums = Array.new(max + 1, 1)
  nums[0] = nums[1] = 0

  (2..Math.sqrt(max)).each do |sieve|
    if nums[sieve] == 1
      (2 * sieve).step(max, sieve).each do |num|
        nums[num] = 0
      end
    end
  end

  nums
end

def prices(max, costs)
  dp = Array.new(max + 1, 0)

  costs.sort!

  for i in 1..max
    costs.each do |cost|
      if i - cost >= 0 && (i - cost == 0 || dp[i - cost] == 1)
        dp[i] = 1
        break
      end
    end
  end

  dp
end

primes = primes(1_000_000)

while line = gets
  n, x = line.chomp.split.map(&:to_i)
  break if n == 0 && x == 0

  costs = []
  n.times do
    costs << gets.chomp.to_i
  end

  prices = prices(x, costs)

  answer = :NA
  x.downto(1) do |num|
    if primes[num] == 1 && prices[num] == 1
      answer = num
      break
    end
  end

  puts answer
end
